File: golang/longest_substring_repeating_characters.go
// brute force
func lengthOfLongestSubstring(s string) int {
	maxLen := 0

	seen := map[byte]bool{}
	for start := 0; start < len(s); start++ {
		seen = map[byte]bool{}
		length := 0
		for end := start; end < len(s); end++ {
			if _, ok := seen[s[end]]; ok {
				if length > maxLen {
					maxLen = length
				}
				break
			}
			seen[s[end]] = true
			length++
		}
	}

	return maxLen
}

// sliding window by char index map
func lengthOfLongestSubstring(s string) int {
	maxLen := 0
	seen := map[byte]int{}
	start, end := 0, 0
	for end < len(s) {
		if index, ok := seen[s[end]]; ok {
			if end-start > maxLen {
				maxLen = end - start
			}

			// remove index map from start to index inclusive
			for start <= index {
				delete(seen, s[start])
				start++
			}
			seen[s[end]] = end

			end++
			continue
		}

		seen[s[end]] = end
		end++
	}
	if end-start > maxLen {
		maxLen = end - start
	}
	return maxLen
}

File: typescript/longest_substring_repeating_characters.ts
function lengthOfLongestSubstring(s: string): number {
	let longest = 0
	let start = 0
	let end = 0

	const indexRef = new Map<string, number>()
	indexRef.set(s[end], 0)

	while (end < s.length) {
		end++

		// found repeating character
		if (indexRef.has(s[end])) {
			longest = Math.max(longest, end - start)

			// delete chars from start to char index inclusive
			let bound = indexRef.get(s[end])
			while (start <= bound) {
				indexRef.delete(s[start])
				start++
			}
		}

		// update window
		indexRef.set(s[end], end)
	}

	// possible that there are never repeats
	longest = Math.max(longest, end - start)

	return longest
}
